Katello/katello
# Fixes #34861 - Syncable yum-format repository exports

#### What are the changes introduced in this pull request?
This commit adds the concept of `syncable` exports. 

```
$ hammer content-export complete version --format=syncable --id=81
[....................................................................................................................................................................] [100%]
Generated /var/lib/pulp/exports/Default_Organization/sample-content-view/4.0/2022-06-30T22-19-34-00-00
```

The files are exported in a different format. cd into `/var/lib/pulp/exports/Default_Organization/sample-content-view/4.0/2022-06-30T22-19-34-00-00` and you should see a directory structure that looks like the following. 
```
$ ls -l /var/lib/pulp/exports/Default_Organization/sample-content-view/4.0/2022-06-30T22-27-11-00-00/content/dist/rhel/server/7/7.5/x86_64/openstack-tools/10/os
total 24300
-rw-rw-rw-+ 1 pulp pulp   19484 Jun 30 22:27 babel-2.3.4-1.el7ost.noarch.rpm
-rw-rw-rw-+ 1 pulp pulp   71976 Jun 30 22:27 openshift-heat-templates-0.9.9-2.el7ost.noarch.rpm
-rw-rw-rw-+ 1 pulp pulp   72448 Jun 30 22:27 openshift-heat-templates-0.9.9-5.el7ost.noarch.rpm
.....
drwxrwxrwx+ 2 pulp pulp    4096 Jun 30 22:27 repodata

$ 
```

The export directory should also display generated listing files (note you need to run hammer as root for this part.)
```
$ find /var/lib/pulp/exports/Default_Organization/sample-content-view/4.0/2022-06-30T22-27-11-00-00/content/ -name listing
....
/var/lib/pulp/exports/Default_Organization/sample-content-view/4.0/2022-06-30T22-27-11-00-00/content/dist/rhel/server/7/listing
/var/lib/pulp/exports/Default_Organization/sample-content-view/4.0/2022-06-30T22-27-11-00-00/content/dist/rhel/server/listing
/var/lib/pulp/exports/Default_Organization/sample-content-view/4.0/2022-06-30T22-27-11-00-00/content/dist/rhel/listing
/var/lib/pulp/exports/Default_Organization/sample-content-view/4.0/2022-06-30T22-27-11-00-00/content/dist/listing
/var/lib/pulp/exports/Default_Organization/sample-content-view/4.0/2022-06-30T22-27-11-00-00/content/listing
....
```

This is similar to the cdn structure. The typical user would
1. Copy the contents of  `/var/lib/pulp/exports/Default_Organization/sample-content-view/4.0/2022-06-30T22-19-34-00-00` to a disconnected server 
2. Update the other server's cdn to `file:///var/lib/pulp/exports/Default_Organization/sample-content-view/4.0/2022-06-30T22-19-34-00-00`
3. Enable and Sync repos available in this archive

#### Constraints
- No `metadata.json` is generated.
- No import command available here. The disconnected user is responsible for the consistency of data.
- No incremental exports. Only complete for starts.

This will be ideally used in environments where one wants to share the exports to the world and the downstream server may be old and not having pulp3 installed. 

#### Considerations taken when implementing this change?
Pulp 3.14 and above had facilities to do a regular complete and incremental exports. However it expects the importer to match versions. In many disconnected environments user may not know what part the downstream may use. They just want a dump of the data like iso's that could then be hand delivered to the disconnected environment.

Pulp 3.16 and later added a new filesystem exports => https://docs.pulpproject.org/pulpcore/restapi.html#tag/Exporters:-Filesystem-Exports
Katello is taking advantage of this feature implement `syncable` exports.

Initially started the PR making this available for repository exports only. However in the current state all three content-export options are available to be extracted in syncable format.
- `hammer content-export complete version --format=syncable ....`
- `hammer content-export complete library --format=syncable ....`
- `hammer content-export complete repository --format=syncable ....`

The export rpm files are generated as hardlinks to the actual file on disk. This there by avoids the 2X space we may need for regular importable exports.  

There are couple of downsides as stated above
- No import command available here. The disconnected user is responsible for the consistency of data.
- No incremental exports. Only complete for starts.

#### What are the testing steps for this pull request?
- Make sure you have https://github.com/Katello/hammer-cli-katello/pull/856 checked out along with this pr.
- On your dev env you may need to run `sudo setfacl --recursive --modify u:vagrant:rwX,d:u:vagrant:rwX /var/lib/pulp/exports` so that listing files can be generate when hammer is running as a vagrant user.
- On your katello instance `Administer -> Settings -> Content`. 
- set both `Default Custom Repository download policy` and `Default Red Hat Repository download policy` to immediate
- Create an org
- Add/Enable + sync few repos (recommend Ansible/Openstack as they tend to be smaller.) 
- Create a content view with these repos an publish.

Try any of the 3 hammer commands listed here
- `hammer content-export complete version --format=syncable ....`
- `hammer content-export complete library --format=syncable ....`
- `hammer content-export complete repository --format=syncable ....`

- Check if the listing files generated ok
